contracts/NFTStorefront.cdc

```diff
@@ -326,7 +405,22 @@ pub contract NFTStorefront {
             // Note that oldOffer will always be nil, but we have to handle it.
             destroy oldOffer
 
-            emit SaleOfferAvailable(saleOfferResourceID: saleOfferResourceID, availableAt: self.owner?.address!)
+            // Format up the cuts in order to include them in the event
+            let cutsRepresentation: {Address: UFix64} = {}
+            for cut in saleCuts {
+                cutsRepresentation[cut.receiver.address] = cut.amount
+            }
+
+            emit SaleOfferAvailable(
+                saleOfferResourceID: saleOfferResourceID,
+                storefrontResourceID: self.uuid,
+                storefrontAddress: self.owner?.address!,
+                nftType: nftType,
+                nftID: nftID,
+                nftCollectionAddress: nftProviderCapability.address,
+                ftVaultType: salePaymentVaultType,
+                cuts: cutsRepresentation
+            )
```

Comment:
Is this event is getting a bit too large? Are all of these values available by querying the Storefront resource?

Eventually events will cost more $$, and I wonder if we should try to promote smaller events paired with a "notify, then fetch" model.

Reply:
There's a tension between "notify and fetch to save gas", "you can reconstitute the complete state of the contract over time from its event log", and "events are useful notifications for off-chain code and should be denormalized to support this". 

This particular event is not the first, is the second, and is mostly the third. 😺

I think it should at least be:
```
    pub event SaleOfferAvailable(
        saleOfferResourceID: UInt64,
        storefrontResourceID: UInt64,
        storefrontAddress: Address
    )
```

As it allows an offchain code to identify the correct source of the offer to fetch more details from. But it does not allow reconstituting the state of the contract, and offchain code will have to parse more events or run some scripts to get all the data they need.

@orodio what do you think?

Reply:
OK after discussion with Dete I'm going with this:
```
    pub event SaleOfferAvailable(
        storefrontAddress: Address,
        saleOfferResourceID: UInt64,
        nftType: Type,
        nftID: UInt64,
        ftVaultType: Type,
        price: UFix64
    )
```
You'll notice that it's three pairs of context/value (where the context is an address or a type).

I've deliberately *not* included the storefront uuid after looking at the NFT standard, where the withdraw and deposit events specify the address but not the uuid:
```
pub event Withdraw(id: UInt64, from: Address?)
```
In the case of the NFT standard this means that if we create several Collections in the same address's storage then link only *one* of them to the well-known public path for that Collection type, and then transfer NFTs between them, we may observe events that we cannot access the tokens of in the expected way.

This is a wart, but I don't want to resolve it here, so I've copied it.

The cuts are also not included. This alone will save a lot of gas.